exercise/static/exercise/chapter.js

```diff
@@ -233,42 +356,204 @@
 						exercise.submit(this);
 					});
 				}
-			}
+			}		
+			
 			$.augmentExerciseGroup(content);
 			window.postMessage({
 				type: "a-plus-bind-exercise",
 				id: this.chapterID
 			}, "*");
 		},
-
-		submit: function(form_element) {
-			//$(form_element).find(":input").prop("disabled", true);
-			//this.showLoader("submit");
-			this.chapter.openModal(this.chapter.messages.submit);
+		
+		// Submit the formData to given url and then execute the callback.
+		submitAjax: function(url, formData, callback, retry) {
 			var exercise = this;
-			$.ajax($(form_element).attr("action"), {
+			$.ajax(url, {
 				type: "POST",
-				data: new FormData(form_element),
+				data: formData,
 				contentType: false,
 				processData: false,
 				dataType: "html"
-			}).fail(function() {
+			}).fail(function(xhr, textStatus, errorThrown) {
+					// handle database lock exceptions
+					retry = retry || 0;
+					if (xhr.responseText.indexOf("database is locked") >= 0 && retry < 5) {
```

Comment:
Is this a special hack for some specific grader? Could it be more generic?

Reply:
Yes, in our test environment using mooc-grader and sqlite multiple grading submissions caused database lock errors for some reason, and this is really just a hack to make the elements usable in that environment. I'm not sure if it is possible to make it more generic since the error code is just 500 and the response text seems be the only thing that contains relevant information of the type of the error. It also anyway works when the Django debug setting is true; otherwise the database lock exception seems to result in 500 status code and the ajax fail parameters don't really have any other useful information to determine what was the cause of the error.

Reply:
Just check that if the server doesn't respond with 200, then retry X times. if server doesn't responding in acceptable time frame, then show error in the exercise.